types/underscore/index.d.ts

```diff
@@ -95,10 +95,31 @@ declare module _ {
         (element: T, key: string, object: V): TResult;
     }
 
-    type CollectionIterator<T, TResult, V> = V extends List<T> ? ListIterator<T, TResult, V>
+    type CollectionIterator<T, TResult, V> =
+        V extends List<T> ? ListIterator<T, TResult, V>
         : V extends Dictionary<T> ? ObjectIterator<T, TResult, V>
         : never;
 
+    type Iteratee<V, R, T extends TypeOfCollection<V> = TypeOfCollection<V>> =
+        undefined |
+        CollectionIterator<T, R, V> |
+        keyof T |
+        string |
+        Array<string | number> |
+        Partial<T>;
+
+    type IterateeResult<V, I, T extends TypeOfCollection<V> = TypeOfCollection<V>> =
+        I extends undefined ? T // default iteratee is _.identity
+        : I extends CollectionIterator<T, infer R, V> ? R
+        : I extends keyof T ? T[I]
+        : I extends string | Array<string | number> ? any
+        : I extends Partial<T> ? boolean
+        : never;
```

Comment:
Exciting to see this implemented!

Reply:
Definitely, thanks for coming up with it!